### PR TITLE
Sentry Functions: Add monitoring

### PR DESCRIPTION
**Pull request by vpandiarajan20**
Add some code to the serverless function so Sentry can monitor the function for errors.

### CODE REVIEW

File: src/sentry/utils/cloudfunctions.py [hunk 1]
```diff
@@ -18,10 +18,45 @@
...
+    return;
+  }
+  if (typeof userFunc !== 'function') {
+    console.error("Your code needs to export a function. Instead, " + typeof userFunc + " was exported.");
+    return;
+  }
+  const event = JSON.parse(Buffer.from(message.data, 'base64').toString());
+  if (process.env.SENTRY_DSN) {
+    const transaction = Sentry.startTransaction({
+        op: "func",
```
Comment (AbhiPrasad) [inline src/sentry/utils/cloudfunctions.py:44]: As per: https://develop.sentry.dev/sdk/performance/span-operations/#serverless-faas

```suggestion
        op: "fass",
```

File: src/sentry/utils/cloudfunctions.py [hunk 2]
```diff
@@ -18,10 +18,45 @@
...
+        name: "Sentry Function Transaction",
```
Comment (AbhiPrasad) [inline src/sentry/utils/cloudfunctions.py:45]: Can we include the function name the user set here?

**Review by AbhiPrasad — COMMENTED** [review]
Quick skim review, on my todo for tomorrow to review properly

File: src/sentry/utils/cloudfunctions.py [hunk 3]
```diff
@@ -18,10 +18,45 @@
...
+    });
+    try {
+        await userFunc(event.data, event.type);
+    } catch (e) {
+        Sentry.captureException(e);
+    } finally {
+        transaction.finish();
```
Comment (lforst) [inline src/sentry/utils/cloudfunctions.py:52]:
```suggestion
        transaction.finish();
        await Sentry.flush(); // Wait until the SDK sent all the events that may have built up
```

File: src/sentry/utils/cloudfunctions.py [hunk 4]
```diff
@@ -18,10 +18,45 @@
 WRAPPER_JS = """
 const userFunc = require('./user.js');
 Object.assign(process.env, require('./env.json'));
+const Sentry = require("@sentry/node");
+const Tracing = require("@sentry/tracing");
+exports.start = async (message, context) => {
+  if (process.env.SENTRY_DSN) {
+    Sentry.init({
```
Comment (lforst) [inline src/sentry/utils/cloudfunctions.py:25]: Important: Just a heads up, make sure `Sentry.init()` is not called within the function/handler itself but in the global scope of a GCP function file.

The reason is that if a function is hot when it gets hit, the SDK is initialized multiple times in the runtime - once for the time the function was cold + any number of additional times for when the function was hot. Multiple init calls in the same global context cause all kinds of breakage and slowdown (stuff being wrapped over and over, breadcrumbs being recorded twice, ...).

File: src/sentry/utils/cloudfunctions.py [hunk 5]
```diff
@@ -18,10 +18,45 @@
...
+      ],
+      tracesSampleRate: 1.0,
+    });
+  }
+  if (!userFunc) {
+    console.error("Your code needs to export a function. module.export = () => {}");
+    return;
+  }
+  if (typeof userFunc !== 'function') {
+    console.error("Your code needs to export a function. Instead, " + typeof userFunc + " was exported.");
```
Comment (markstory) [inline src/sentry/utils/cloudfunctions.py:38]:
```suggestion
    console.error("Your code needs to export a function. Instead, " + typeof userFunc + " was exported. Make sure your code contains an export like `module.export = () => { /* your code here */ }`");
```

We might want to spell out what is required more clearly.

File: src/sentry/utils/cloudfunctions.py [hunk 6]
```diff
@@ -18,51 +18,39 @@
...
     try {
-        await userFunc(event.data, event.type);
+      userFunc(event.data, event.type);
     } catch (e) {
-        Sentry.captureException(e);
+      Sentry.captureException(e);
     } finally {
-        transaction.finish();
+      transaction.finish();
+      Sentry.flush();
```
Comment (AbhiPrasad) [inline src/sentry/utils/cloudfunctions.py:43]: All of this shouldn't be needed, the `wrapCloudEventFunction` func should be handling this logic (Creating a new transaction, automatically catching errors, automatically flushing from the queue): https://github.com/getsentry/sentry-javascript/blob/549370c7709aa96df4608b2b204bf06e23cd67ea/packages/serverless/src/gcpfunction/cloud_events.ts#L24

**Pull request closed without merging**